src/main/java/com/cedarsoftware/util/UsageTrackingMap.java

```diff
@@ -0,0 +1,101 @@
+package com.cedarsoftware.util;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class UsageTrackingMap<K, V> implements Map<K, V> {
+    private final Map<K, V> internalMap;
+    private Set<Object> readKeys;
+
+    public UsageTrackingMap(Map<K, V> map) {
+        internalMap = map; //TODO What to do when input map is null?
+        readKeys = new HashSet<>();
+    }
+
+    public V get(Object key) {
+        V value = internalMap.get(key);
+        if (value != null) {
+            readKeys.add(key);
+        }
+        return value;
+    }
+
+    public V put(K key, V value) {
+        return internalMap.put(key, value);
+    }
+
+    public boolean containsKey(Object key) {
+        boolean containsKey = internalMap.containsKey(key);
+        if (containsKey) {
+            readKeys.add(key);
+        }
+        return containsKey;
+    }
+
+    public void putAll(Map<? extends K, ? extends V> m) {
+        internalMap.putAll(m);
+    }
+
+    public V remove(Object key) {
+        readKeys.remove(key);
+        return internalMap.remove(key);
+    }
+
+    public int size() {
+        return internalMap.size();
+    }
+
+    public boolean isEmpty() {
+        return internalMap.isEmpty();
+    }
+
+    public boolean equals(Object other) {
+        return other instanceof UsageTrackingMap && internalMap.equals(((UsageTrackingMap) other).internalMap);
```

Comment:
I believe we want equality between Maps to only consider Mapped contents, not the Map instance itself.  If I have a ConcurrentHashMap and a LinkedHashMap with the same contents, are they .equals()?  I believe so.  Bringing in the pull request, regardless, would like to discuss further.